### PR TITLE
drivers: gpio: shell: Improved error reporting

### PR DESCRIPTION
**Pull request by aasinclair**
Invalid device now generates message and errno return 
Arguments checked with shell_strtoul
gpio function call return values checked and passed on

Device name completion and usage text added

SHELL_GPIO enabled when SHELL and GPIO both enabled

Configuration command now accepts integer argument as well and named modes

### CODE REVIEW

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Looks good, thanks! One minor nit:

File: drivers/gpio/gpio_shell.c [hunk 1]
```diff
@@ -21,86 +21,75 @@
...
-			type = GPIO_INPUT | GPIO_PULL_UP;
-		} else if (!strcmp(argv[args_indx.mode], "ind")) {
-			type = GPIO_INPUT | GPIO_PULL_DOWN;
-		} else if (!strcmp(argv[args_indx.mode], "out")) {
-			type = GPIO_OUTPUT;
-		} else {
-			return 0;
-		}
+	dev = device_get_binding(argv[1]);
+	if (dev == NULL) {
```
[the remark this answers is not in the document]
Comment (aasinclair) [reply drivers/gpio/gpio_shell.c:32]: > Please use `device_is_ready()` instead of just a NULL check.

Hi, looking at the implementation for device_get_binding, it already uses device_is_ready internally, so I think the NULL check is correct.

**Review by henrikbrixandersen — DISMISSED** [review]

File: drivers/gpio/gpio_shell.c [hunk 2]
```diff
@@ -21,86 +21,75 @@
...
-		} else if (!strcmp(argv[args_indx.mode], "inu")) {
-			type = GPIO_INPUT | GPIO_PULL_UP;
-		} else if (!strcmp(argv[args_indx.mode], "ind")) {
-			type = GPIO_INPUT | GPIO_PULL_DOWN;
-		} else if (!strcmp(argv[args_indx.mode], "out")) {
-			type = GPIO_OUTPUT;
-		} else {
-			return 0;
-		}
+	dev = device_get_binding(argv[1]);
```
Comment (WalkingTalkingPotato) [inline drivers/gpio/gpio_shell.c:31]: I think it would be better for readability to make local variables for arguments instead of using arbitrary array elements directly, something like this:

`char *port = argv[1];

...

dev = device_get_binding(port);`

File: drivers/gpio/gpio_shell.c [hunk 3]
```diff
@@ -21,86 +21,75 @@
...
+		type = GPIO_INPUT | GPIO_PULL_UP;
+	} else if (!strcmp(argv[3], "ind")) {
+		type = GPIO_INPUT | GPIO_PULL_DOWN;
+	} else if (!strcmp(argv[3], "out")) {
+		type = GPIO_OUTPUT;
 	} else {
+		type = shell_strtoul(argv[3], 0, &rc);
+	}
+
+	index = shell_strtoul(argv[2], 10, &rc);
```
Comment (WalkingTalkingPotato) [inline drivers/gpio/gpio_shell.c:49]: index is uint8_t, so result of shell_strtoul() needs to be checked for overflow.

File: drivers/gpio/gpio_shell.c [hunk 4]
```diff
@@ -21,86 +21,75 @@
...
-	.port = 1,
-	.index = 2,
-	.mode = 3,
-	.value = 3,
-};
-
 static int cmd_gpio_conf(const struct shell *sh, size_t argc, char **argv)
 {
 	uint8_t index = 0U;
 	int type = GPIO_OUTPUT;
```
Comment (WalkingTalkingPotato) [inline drivers/gpio/gpio_shell.c:27]: No need for initial value; can be moved closer to usage

File: drivers/gpio/gpio_shell.c [hunk 5]
```diff
@@ -21,86 +21,75 @@
...
-static const struct args_index args_indx = {
-	.port = 1,
-	.index = 2,
-	.mode = 3,
-	.value = 3,
-};
-
 static int cmd_gpio_conf(const struct shell *sh, size_t argc, char **argv)
 {
 	uint8_t index = 0U;
```
Comment (WalkingTalkingPotato) [inline drivers/gpio/gpio_shell.c:26]: Please consider using gpio_pin_t type;
Can be moved closer to usage

File: drivers/gpio/gpio_shell.c [hunk 6]
```diff
@@ -21,86 +21,75 @@
...
+
+	index = shell_strtoul(argv[2], 10, &rc);
+	if (rc < 0) {
 		shell_error(sh, "Wrong parameters for conf");
-		return -ENOTSUP;
+		return rc;
 	}
 
-	dev = device_get_binding(argv[args_indx.port]);
+	shell_print(sh, "Configuring %s pin %d with mask %d", argv[1], index, type);
```
Comment (WalkingTalkingPotato) [inline drivers/gpio/gpio_shell.c:55]: %d expects signed int argument, but "type" is uint8_t - could that lead to undefined behavior?

File: drivers/gpio/gpio_shell.c [hunk 7]
```diff
@@ -21,86 +21,75 @@
...
+	if (!strcmp(argv[3], "in")) {
+		type = GPIO_INPUT;
+	} else if (!strcmp(argv[3], "inu")) {
+		type = GPIO_INPUT | GPIO_PULL_UP;
+	} else if (!strcmp(argv[3], "ind")) {
+		type = GPIO_INPUT | GPIO_PULL_DOWN;
+	} else if (!strcmp(argv[3], "out")) {
+		type = GPIO_OUTPUT;
 	} else {
+		type = shell_strtoul(argv[3], 0, &rc);
```
Comment (WalkingTalkingPotato) [inline drivers/gpio/gpio_shell.c:46]: rc value not checked

File: drivers/gpio/gpio_shell.c [hunk 8]
```diff
@@ -143,45 +137,74 @@ static int cmd_gpio_blink(const struct shell *sh,
...
+}
+
+SHELL_DYNAMIC_CMD_CREATE(dsub_device_name, device_name_get);
+
+SHELL_STATIC_SUBCMD_SET_CREATE(
+	sub_gpio,
+	SHELL_CMD_ARG(conf, &dsub_device_name,
+		     "Configure GPIO\n"
+		     "Usage: conf <device> <pin> <mode>\n"
+		     "The mode argument can be an integer, or one of:\n"
```
Comment (WalkingTalkingPotato) [inline drivers/gpio/gpio_shell.c:192]: I think it's better to rephrase it to emphasize default "premade" values and custom nature of integer, something like this:

Mode argument accepts these values:
1. in - Input (floating)
2. inu - Input (with pull-up)
3. ind - Input (with pull-down)
4. out - Output (push-pull)
5. <any integer> - Custom mode value passed directly to gpio_pin_configure() flags argument.

File: drivers/gpio/gpio_shell.c (hunk 6 above)
Comment (WalkingTalkingPotato) [inline drivers/gpio/gpio_shell.c:55]: %d expects signed int argument, but "type" is uint8_t - could that lead to undefined behavior?

File: drivers/gpio/gpio_shell.c (hunk 3 above)
Comment (WalkingTalkingPotato) [inline drivers/gpio/gpio_shell.c:49]: index is uint8_t, so result of shell_strtoul() needs to be checked for overflow.

File: drivers/gpio/gpio_shell.c (hunk 2 above)
Comment (WalkingTalkingPotato) [inline drivers/gpio/gpio_shell.c:31]: I think it would be better for readability to make local variables for arguments instead of using arbitrary array elements directly, something like this:

`char *port = argv[1];

...

dev = device_get_binding(port);`

File: drivers/gpio/gpio_shell.c (hunk 4 above)
Comment (WalkingTalkingPotato) [inline drivers/gpio/gpio_shell.c:27]: No need for initial value;
Can be moved closer to usage;
gpio_pin_configure() expects gpio_flags_t type (typedef for uint32_t), not int

File: drivers/gpio/gpio_shell.c (hunk 5 above)
Comment (WalkingTalkingPotato) [inline drivers/gpio/gpio_shell.c:26]: Please consider using gpio_pin_t type;
Can be moved closer to usage

File: drivers/gpio/gpio_shell.c (hunk 6 above)
Comment (aasinclair) [reply drivers/gpio/gpio_shell.c:55]: This is OK, it will be implicitly casted.
It's only an issue when using pointers (e.g. sscanf).

File: drivers/gpio/gpio_shell.c (hunk 3 above)
Comment (aasinclair) [reply drivers/gpio/gpio_shell.c:49]: The types are pretty untidy, I'll update them all to unsigned long to match strtoul return value.

File: drivers/gpio/gpio_shell.c (hunk 7 above)
Comment (aasinclair) [reply drivers/gpio/gpio_shell.c:46]: it is checked further down, as shell_strtoul can be stacked (existing pointer to error is not overwritten)

Comment (WalkingTalkingPotato) [reply drivers/gpio/gpio_shell.c:46]: No, it's overwritten at line 49:
'index = shell_strtoul(argv[2], 10, &rc);'

Comment (WalkingTalkingPotato) [reply drivers/gpio/gpio_shell.c:46]: No, it's overwritten at line 49:
'index = shell_strtoul(argv[2], 10, &rc);'

**Review by WalkingTalkingPotato — APPROVED** [review]

File: drivers/gpio/Kconfig [hunk 9]
```diff
@@ -16,6 +16,7 @@ source "subsys/logging/Kconfig.template.log_config"
 
 config GPIO_SHELL
 	bool "GPIO Shell"
+	default y
```
Comment (carlescufi) [inline drivers/gpio/Kconfig:19]: Why?

Comment (aasinclair) [reply drivers/gpio/Kconfig:19]: This way you get GPIO shell automatically if you have GPIO and SHELL enabled.
Otherwise you have to enable it explicitly.

Most of the other shell implementations have done it this way.

**Review by gmarull — APPROVED** [review]

**Comment by nixward** [comment]
Hi @aasinclair, apologies up front from me, I forgot to check if anyone else had been working on this shell and ended up recently putting together a PR for it also see [PR 63019](https://github.com/zephyrproject-rtos/zephyr/pull/63019/files). The lack of tab suggestion and completion functionality was the bad user experience I'd recently had which was the trigger for me to try to find a solution. In the PR I used the dynamic subcommand and the dictionary subcommand APIs to enable tab suggestion/complete feature for all gpio_shell subcommands. The benefit of re-structuring the gpio_shell to utilise the dynamic subcommand API for device name and pin sub commands, and the dictionary subcommand API for pin config options and pin set values is that if the user manually enters something not in the subcommand set the shell will automatically reject it and will print out the valid subcommand options available _which removes the need to write error checking code for these subcommands_ as it's already built into these subcommand APIs.

In addition I can now see there's a way forward with this tab suggestion/completion strategy to use the Devicetree to limit the device suggestions to only gpio devices similar to the adc_shell ([link](https://github.com/zephyrproject-rtos/zephyr/blob/73b803ab4b50847f3865f3da1b175b07038a9f62/drivers/adc/adc_shell.c#L72-L107)) and to use the Devicetree to limit the pin subcommands to only those indicated as available by these [gpio bindings properties](https://github.com/zephyrproject-rtos/zephyr/blob/73b803ab4b50847f3865f3da1b175b07038a9f62/dts/bindings/gpio/gpio-controller.yaml#L15-L37).

So in summary I think tab suggestion/completion functionality in the gpio_shell would be a good base line for it to have so I implore you and others to have a look at [PR 63019](https://github.com/zephyrproject-rtos/zephyr/pull/63019/files) as an alternate way forward. If there's consensus PR 63019 is a desirable way forward I'm of course happy to modify with ideas from this PR or if alternatively this PR is deemed the preferred way feel free to pull in any ideas from PR 63019. Thanks.

**Comment by aasinclair** [comment]
> Hi @aasinclair, apologies up front from me, I forgot to check if anyone else had been working on this shell and ended up recently putting together a PR for it also see [PR 63019](https://github.com/zephyrproject-rtos/zephyr/pull/63019). The lack of tab suggestion and completion functionality was the bad user experience I'd recently had which was the trigger for me to try to find a solution. In the PR I used the dynamic subcommand and the dictionary subcommand APIs to enable tab suggestion/complete feature for all gpio_shell subcommands. The benefit of re-structuring the gpio_shell to utilise the dynamic subcommand API for device name and pin sub commands, and the dictionary subcommand API for pin config options and pin set values is that if the user manually enters something not in the subcommand set the shell will automatically reject it and will print out the valid subcommand options available _which removes the need to write error checking code for these subcommands_ as it's already built into these subcommand APIs.
> 
> In addition I can now see there's a way forward with this tab suggestion/completion strategy to use the Devicetree to limit the device suggestions to only gpio devices similar to the adc_shell ([link](https://github.com/zephyrproject-rtos/zephyr/blob/73b803ab4b50847f3865f3da1b175b07038a9f62/drivers/adc/adc_shell.c#L72-L107)) and to use the Devicetree to limit the pin subcommands to only those indicated as available by these [gpio bindings properties](https://github.com/zephyrproject-rtos/zephyr/blob/73b803ab4b50847f3865f3da1b175b07038a9f62/dts/bindings/gpio/gpio-controller.yaml#L15-L37).
> 
> So in summary I think tab suggestion/completion functionality in the gpio_shell would be a good base line for it to have so I implore you and others to have a look at [PR 63019](https://github.com/zephyrproject-rtos/zephyr/pull/63019) as an alternate way forward. If there's consensus PR 63019 is a desirable way forward I'm of course happy to modify with ideas from this PR or if alternatively this PR is deemed the preferred way feel free to pull in any ideas from PR 63019. Thanks.

Hi @nixward.
I've just tested your PR and the autocompletion is great, so I think it would be prudent to take your PR as a starting point.
The only features I would really like to see from this PR are:
* Add default y to the Kconfig, so you don't have to manually specify CONFIG_GPIO_SHELL (Just CONFIG_SHELL is enough).
* Add the option for custom configuration of the pin (in addition to in, inu, ind, out).

**Comment by WalkingTalkingPotato** [comment]
In theory it is possible that an API call will return an error even if all the arguments were correct. Example - if we're controlling a GPIO extender, connected through an external bus (say SPI or I2C), call to it could fail if the bus connection got disrupted. So, I disagree that autocompletion "removes the need to write error checking code".

**Comment by nixward** [comment]
@WalkingTalkingPotato yes I agree, sorry I should have been more specific, it removes the need for error handling of some incorrect input, i.e. the gpio_pin_set() value will always be 0 or 1 as that is all the subcommand will allow otherwise the dictionary subcommand API will inform the user they need to select 0 or 1. But yes if gpio_pin_set() returns an error it should be reported via the shell so I'll double check that is handled. Please feel free to point out anything I've missed.
